autoemulate/experimental/emulators/gaussian_process/exact.py

```diff
@@ -0,0 +1,200 @@
+import logging
+
+import gpytorch
+import numpy as np
+import torch
+from gpytorch import ExactMarginalLogLikelihood
+from gpytorch.distributions import MultitaskMultivariateNormal, MultivariateNormal
+from gpytorch.likelihoods import MultitaskGaussianLikelihood
+from gpytorch.kernels import (
+    ScaleKernel,
+)
+from torch import nn
+
+from autoemulate.experimental.data.preprocessors import Preprocessor, Standardizer
+from autoemulate.experimental.emulators.base import (
+    Emulator,
+    InputTypeMixin,
+)
+from autoemulate.emulators.gaussian_process import (
+    zero_mean,
+    constant_mean,
+    linear_mean,
+    poly_mean,
+)
+from autoemulate.emulators.gaussian_process import (
+    rbf,
+    matern_5_2_kernel,
+    matern_3_2_kernel,
+    rq_kernel,
+    rbf_plus_constant,
+    rbf_plus_linear,
+    matern_5_2_plus_rq,
+    rbf_times_linear,
+)
+from autoemulate.experimental.emulators.gaussian_process import (
+    CovarModuleFn,
+    MeanModuleFn,
+)
+from autoemulate.experimental.types import InputLike, OutputLike
+from autoemulate.utils import set_random_seed
+
+
+class GaussianProcessExact(
+    Emulator, InputTypeMixin, gpytorch.models.ExactGP, Preprocessor
+):
+    """
+    Gaussian Process Exact Emulator
+    This class implements an exact Gaussian Process emulator using the GPyTorch library
+
+    It supports:
+    - multi-task Gaussian processes
+    - custom mean and kernel specification
+
+    """
+
+    def __init__(
+        self,
+        x: InputLike,
+        y: InputLike,
+        likelihood_cls: type[MultitaskGaussianLikelihood],
+        mean_module_fn: MeanModuleFn,
+        covar_module_fn: CovarModuleFn,
+        preprocessor_cls: type[Preprocessor] | None = None,
+        random_state: int | None = None,
+        epochs: int = 10,
+        batch_size: int = 16,
+        activation: type[nn.Module] = nn.ReLU,
+        lr: float = 0.01,
+    ):
+        if random_state is not None:
+            set_random_seed(random_state)
+
+        x, y = self._convert_to_tensors(x, y)
+
+        # Initialize the mean and covariance modules
+        # TODO: consider refactoring to only pass torch tensors x and y
+        mean_module = mean_module_fn(tuple(x.shape)[1], torch.Size([y.shape[1]]))
+        combined_kernel = covar_module_fn(tuple(x.shape)[1], torch.Size([y.shape[1]]))
```

Comment:
This is super minor but it doesn't look like we are combining kernels here so can we call it something like `covar_module` rather than `combined_kernel` (unless I missed something)?

Reply:
Yes agreed, `covar_module` sounds good so it would be consistent with `mean_module` too.

Reply:
Added in [9551465](https://github.com/alan-turing-institute/autoemulate/pull/384/commits/9551465ee53039cfaaaedabbd12f330f4b99b9b5)